### ISSUE TITLE
CudnnRnnSequenceTensorDescriptor should support different sequence lengths

### ISSUE THREAD

**Issue by albertz** [tensorflow/tensorflow#6633]
According to the cuDNN docs, the functions `cudnnRNNForwardInference` / `cudnnRNNForwardTraining` get the argument `cudnnTensorDescriptor_t* xDesc`, where:

xDesc: Array of tensor descriptors. Each must have the same second dimension. The first dimension may decrease from element n to element n + 1 but may not increase.

The usage of `xDesc` is a bit non-straight-forward. I wrote about that in more detail [here](http://stackoverflow.com/questions/41461670/cudnnrnnforwardtraining-seqlength-xdesc-usage).
According to a [comment in the CNTK code](https://github.com/Microsoft/CNTK/blob/7c5fb2d7d806148b5cbd795407f7c7b6a1a64520/Source/Math/CuDnnRNN.cpp) about the dimensions of each `xDesc[t]`:

> these dimensions are what CUDNN expects: (the minibatch dimension, the data dimension, and the number 1 (because each descriptor describes one frame of data)

TensorFlow sets the same minibatch dimension for each `xDesc[t]` in `CudnnRnnSequenceTensorDescriptor`:

    int dims[] = {batch_size, data_size, 1};
    int strides[] = {dims[1] * dims[2], dims[2], 1};
    status = dynload::cudnnSetTensorNdDescriptor(
        parent, handle /*tensorDesc*/, data_type /*dataType*/,
        sizeof(dims) / sizeof(dims[0]) /*nbDims*/, dims /*dimA*/,
        strides /*strideA*/);
    CUDNN_RETURN_IF_FAIL(status, "Failed to update tensor descriptor");
    // Replicate handle across the number of steps.
    handles_.assign(seq_length, handle);

Also `createRnnSequenceTensorDescriptor` needs a new API to allow for that.

And I'm not sure if there are ways to prepare the input `x` for `cudnnRNNForwardTraining` so that it has all sequences contiguously behind each other, and the sequences are sorted by sequences length.
Similar as [`PackSequencesForCuDNN` in CNTK](https://github.com/Microsoft/CNTK/blob/4472649412929543d4dfe553f50be5d9b3102521/Source/ComputationNetworkLib/RNNNodes.cpp#L265).

**Comment by alquraishi** [comment]
Are there any updates on this? And how does it relate to this [issue](https://groups.google.com/a/tensorflow.org/forum/#!searchin/discuss/alquraishi/discuss/Q4egUFzG_3I/IZ5NMQIxAQAJ)?

**Comment by zheng-xq** [comment]
Assigning James who works on cudnn-rnn extensively.

**Comment by protoget** [comment]
@albertz 
I think it's actually more of a question to Nvidia than to TF.
1. I'm not sure why we chose 3 dimension, for the TensorNdDescriptor -- but from Cudnn 7.0 it actually suggests at least 4 dims:
> The total size of a tensor including the potential padding between dimensions is
limited to 2 Giga-elements of type datatype . Tensors are restricted to having at least
4 dimensions, and at most CUDNN_DIM_MAX dimensions (defined in cudnn.h). When
working with lower dimensional data, it is recommended that the user create a 4D
tensor, and set the size along unused dimensions to 1.

2. We don't yet have functions to help doing that. The sorting and packing (and later unsorting/unpacking) is very expensive.

**Comment by alquraishi** [comment]
@protoget it appears that Keras does support this however, according to this [post](https://groups.google.com/a/tensorflow.org/forum/#!searchin/discuss/alquraishi/discuss/Q4egUFzG_3I/IZ5NMQIxAQAJ)?

Irrespective of the padding issue, if the input tensor changes in size (say due to bucketing so that from batch to batch the size of the input tensor is different), would the ops and layers defined in `cudnn_rnn.py` function correctly or not? I understand that they don't currently account for different sequence lengths within a batch, but are they at least able to deal with dynamically sized input tensors?

**Comment by albertz** [comment]
You can use [`tf.reverse_sequence`](https://www.tensorflow.org/api_docs/python/tf/reverse_sequence) to reverse it for the backward direction. Then you can reverse it afterwards again. You have to do it separately for the forward and backward direction, and also separately for each layer. I implemented that in our framework [Returnn](https://github.com/rwth-i6/returnn). Of course, that degrades the performance. I did a benchmark [here](http://returnn.readthedocs.io/en/latest/tf_lstm_benchmark.html).

**Comment by protoget** [comment]
@alquraishi

Keras has support of padding with a performance hit. I don't know if it's better/worst than if implemented using CuDNN's native support (which is also expensive).

inputs to cudnn_rnn classes have 3 dims: [time_steps, batch_size, input_dim]. The 3rd dim is a constant which doesn't change across batches. It's a requirement for all LSTM impl, regardless of what devices they are.

**Comment by alquraishi** [comment]
@protoget thanks for the clarification about the 3 dims. My question was with regards to time_steps. Is it ok if that changes from batch to batch? I have it running and it's not throwing up an error, but I want to make sure the resulting semantics are correct (modulo padding issue--I understand that the padded time steps are treated as actual entries.)

**Comment by protoget** [comment]
It's ok if the sequence_length varies from batch to batch.

**Comment by protoget** [comment]
We'll probably support variable sequence length the same way Keras does. I don't seem to have the bandwidth in recent 2-3 weeks however. Will follow up after that.

**Comment by alquraishi** [comment]
One question / comment about supporting the correct zero-padding semantics for variable lengths within a batch. It appears to me that if one assumes all zeros for the initial hidden and cell states, which is the default behavior (the user must explicitly pass in an initial state tensor to change this), then the fix to get the correct semantics is rather simple. That's because, looking at the math for LSTMs, if both the inputs and previous states are zero, then the output of an LSTM cell is also zero. And so as long as the initial state is zero for the backward direction, then all the padding zeros encountered before the real sequence begins will simply be ignored. The only issue is that the forward direction will carry on some residual signal even after it encounters the zero padding. Thus, the solution would be to zero out all the outputs of the LSTM that correspond to the zero padding regions, before feeding these outputs to the next layer in a multi-layer stack. This precludes using the native cuDNN functionality for multiple layers, but AFAICT there's no speed difference for _bidirectional_ LSTMs between manually constructing multiple layers, and using the native functionality in cuDNN. So this seems like a much simpler and cheaper solution than using `tf.reverse_sequence`, etc.

**Comment by protoget** [comment]
@alquraishi 

The following conclusion isn't correct:
> if both the inputs and previous states are zero, then the output of an LSTM cell is also zero
Each gate in CuDNN LSTM (which isn't the same as LSTM from wiki) has two bias terms -- which means with zero input and state, you get zero output/new state only if they're zero.

As for:
>  but AFAICT there's no speed difference for bidirectional LSTMs between manually constructing multiple layers, and using the native functionality in cuDNN.
I don't think so. CuDNN fused bidirectional multi-layer impl is faster than if manually stacking concatenated outputs from each layer.

**Comment by alquraishi** [comment]
Yes you're right, I didn't think about the bias terms. I did some numerical tests and the outputs came out identical, but that must be because the bias terms were small enough that things cancelled out. That's too bad :(

Regarding speed, in the tests I've done I did not find a detectable difference between manual stacking and CuDNN fused layers. It's possible that other parts of the model are adding sufficient overhead to mask the difference, but it also makes sense that it wouldn't make a huge difference, since both directions have to be integrated before the next layer can begin to be computed, and so there's no real extra parallelism that can be exploited, and presumably whatever overhead is added due to taking outputs from one layer and feeding them into the next is negligible for reasonably large LSTMs. I'm running with ~1000 units per direction and hundreds of time steps, and the slowdown, if any, is less than 1%.

Given that my idea is wrong... any chance this'll be implemented soon or should I look into splitting directions and using `tf.reverse_sequence`?

Sorry about the false alarm.

**Comment by alquraishi** [comment]
I went ahead and implemented a corrected (hopefully for real this time) version by splitting the two directions and then using `tf.reverse_sequence` on the inputs and outputs of the backward direction. This appears to work correctly with more sensitive tests that I now have that fail with the previous incorrect approach. Surprisingly, the performance penalty appears non-existent! On a fairly bare bones model with two bidirectional layers with 800 units for each direction, and a maximum sequence length of around 1,000, I'm seeing absolutely no difference in speed between the fused bidirectional cuDNN op and this split approach (in fact the latter is 5% faster, but that's likely noise). This is on a system with Titan X Pascal, running on TF 1.4, Ubuntu 14.04, CUDA 8, and cuDNN 7 (compiled from source.) The tests were done on a fully loaded system, and are an average of separate 10 runs, each of which consumed 100 batches, so the numbers should be trustworthy.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by alextp** [comment]
I think @alquraishi 's suggestion solves the problem. Please reopen if it doesn't.

**Comment by albertz** [comment]
No, it does not. It describes a slow workaround which everyone is doing now because this is not correctly implemented.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jm565** [comment]
Any updates on this?

@alquraishi Sounds like you had no problems with the workaround. Could you elaborate with a pseudo code snippet how to make it work?

When using two unidirectional CudnnLSTMs and using tf.reverse_sequence on the inputs for the backward direction and on its outputs, do i still have to manually mask the outputs or just concatenate them?

**Comment by bstriner** [comment]
Hi everybody! Just implemented variable sequence length CuDNN RNNs, which touches on this subject. PR here: https://github.com/tensorflow/tensorflow/pull/22308

Cheers

**Comment by reuben** [comment]
I think this was fixed by #23588?

**Comment by dynamicwebpaige** [comment]
Closing this issue, as the capability was introduced in PR #23588.
